File: key-system/generate.py
import string, hashlib, json, os

SECRET = "IRISH_HUB_2024_X9K2"
CHARSET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
CSET_LEN = 36

def crc32_custom(data):
    crc = 0xFFFFFFFF
    for b in data:
        byte = ord(b) if isinstance(b, str) else b
        crc ^= byte
        for _ in range(8):
            if crc & 1:
                crc = (crc >> 1) ^ 0xEDB88320
            else:
                crc >>= 1
    return crc ^ 0xFFFFFFFF

def make_checksum(payload_8chars):
    data = SECRET + payload_8chars
    h = crc32_custom(data)
    s = ""
    for _ in range(4):
        s += CHARSET[h % CSET_LEN]
        h //= CSET_LEN
    return s

def generate_key(index):
    h = hashlib.sha256(f"{SECRET}:{index}".encode()).hexdigest()
    payload = ""
    for i in range(8):
        payload += CHARSET[int(h[i * 2:(i + 1) * 2], 16) % CSET_LEN]
    check = make_checksum(payload)
    return f"IRISH-{payload[:4]}-{payload[4:8]}-{check}"

def validate_key(key):
    if not key.startswith("IRISH-"):
        return False
    parts = key.split("-")
    if len(parts) != 4:
        return False
    for p in parts[1:]:
        if len(p) != 4:
            return False
        for c in p:
            if c not in CHARSET:
                return False
    payload = parts[1] + parts[2]
    expected = make_checksum(payload)
    return parts[3] == expected

keys = [generate_key(i) for i in range(500)]

print(f"Generated {len(keys)} keys")
print(f"First: {keys[0]}, Validate: {validate_key(keys[0])}")
print(f"Fake: {validate_key('IRISH-AAAA-BBBB-CCCC')}")

with open(r"C:\Users\aerux\OneDrive\Documents\Default Project\key-system\keys.json", "w") as f:
    json.dump(keys, f)
print("Saved keys.json")

File: key-system/generate.js
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// SECRET - only you know this, not in the script
const SECRET = "IRISH_HUB_SECRET_2024_X9K2";
const NUM_KEYS = 500;
const CHARSET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

function generateKey(index) {
    const hash = crypto.createHmac('sha256', SECRET).update(String(index)).digest('hex');
    // Take 12 chars from hash, map to charset (3 groups of 4)
    let key = "IRISH-";
    for (let i = 0; i < 12; i++) {
        if (i > 0 && i % 4 === 0) key += "-";
        const idx = parseInt(hash[i], 16) % CHARSET.length;
        key += CHARSET[idx];
    }
    return key;
}

function validateKey(key) {
    // Format check
    if (!/^IRISH-[A-Z0-9]{4}-[A-Z0-9]{4}-[A-Z0-9]{4}$/.test(key)) return false;
    // Brute check: regenerate all and compare
    for (let i = 0; i < NUM_KEYS; i++) {
        if (generateKey(i) === key) return true;
    }
    return false;
}

// Generate all keys
const keys = [];
for (let i = 0; i < NUM_KEYS; i++) {
    keys.push(generateKey(i));
}

// Save keys list
fs.writeFileSync(path.join(__dirname, 'keys.txt'), keys.join('\n'));

// Save keys as compact JS array for web page
const keysJS = `const VALID_KEYS = ${JSON.stringify(keys)};\n`;
fs.writeFileSync(path.join(__dirname, 'keys_data.js'), keysJS);

console.log(`Generated ${NUM_KEYS} keys`);
console.log(`First 5: ${keys.slice(0, 5).join(', ')}`);
console.log(`Last: ${keys[keys.length - 1]}`);
console.log(`Validate test: ${validateKey(keys[0])} (should be true)`);
console.log(`Validate test: ${validateKey('IRISH-AAAA-BBBB-CCCC')} (should be false)`);
